vignettes/readr.Rmd

```diff
@@ -142,15 +142,6 @@ mtcars_spec
 cols_condense(mtcars_spec)
 ```
 
-
-By default readr only looks at the first 1000 rows. This keeps file parsing speedy, but can generate incorrect guesses. For example, in `challenge.csv` the column types change in row 1001, so readr guesses the wrong types. One way to resolve the problem is to increase the number of rows:
-
```

Comment:
This is no longer true so it's been removed. Now, readr (via vroom) selects semi-randomized rows to select for guess, and always includes the last row for guessing. So increasing `guess_max` won't necessarily resolve problems in this case.

https://github.com/tidyverse/vroom/blob/548344f4826818ca988e80b90cdf7fd837a90561/src/collectors.h#L161-L189

Reply:
The rows are not selected in a random fashion. They're basically evenly spaced within the file and the last row is always included for guessing.

https://github.com/tidyverse/vroom/blob/0e2df4cc8a07fbb007206f6bfcc6a3732e6c4613/src/collectors.h#L164

Reply:
I know, but I felt like being overly specific would be more confusing. I can modify to say "interspersed through-out the file and always includes the first and last row" instead.

Reply:
Actually, this isn't technically true since `guess_max` doesn't guess column types using rows it will never read and once we complete https://github.com/tidyverse/readr/pull/1436 then `n_max = guess_max`. So if `guess_max = 1500` then `n_max = 1500` and it will guess using the first 1500 rows in the data. So `spec_*()` guessing will never be interspersed throughout the file, it'll always use all of the data it is given to guess. Here is an example using the original `challenge.csv` where the column `y` is filled with `NA` until row 1001.

``` r
# after we complete #1436
spec_csv(readr_example("challenge.csv"), guess_max = 1000) # n_max = 1000
#> cols(
#>   x = col_double(),
#>   y = col_logical()
#> )
# increasing guess_max does change the spec
spec_csv(readr_example("challenge.csv"), guess_max = 1001) # n_max = 1001
#> cols(
#>   x = col_double(),
#>   y = col_date(format = "")
#> )

# compared to how guessing works in read_csv
# since it's dispersed throughout the file, it guesses correctly
spec(read_csv(readr_example("challenge.csv"), guess_max = 1000, show_col_types = FALSE)) # n_max = all the data
#> cols(
#>   x = col_double(),
#>   y = col_date(format = "")
#> )
```

Reply:
The statement should not be technically misleading (e.g. suggesting that it's random and might select different rows each time you call it), but otherwise it's fine to stay high-level. So, yes, something along these lines: "interspersed through-out the file and always includes the first and last row".

Recall that we did create an entire new vignette on column type guessing, in case it makes sense to link to that (or borrow test from that), for more details: https://readr.tidyverse.org/articles/column-types.html.

To zoom out, I think a guiding principle is to avoid making this text so specific that it's tightly coupled with detail work like #1436. What level of understanding and detail is useful for the typical user?

The observation above about `n_max` and `guess_max` in `spec()` brings up an important question to think about over in #1436. Should `spec_*()` produce the same column type guesses as `read_*()` at default settings?